GPU/GPUCommonHW.cpp

```diff
@@ -831,24 +831,36 @@ void GPUCommonHW::Execute_VertexType(u32 op, u32 diff) {
 }
 
 void GPUCommonHW::Execute_VertexTypeSkinning(u32 op, u32 diff) {
-	// Don't flush when weight count changes.
+	// Don't flush when only weight count changes.
 	if (diff & ~GE_VTYPE_WEIGHTCOUNT_MASK) {
-		// Restore and flush
-		gstate.vertType ^= diff;
-		Flush();
-		gstate.vertType ^= diff;
-		if (diff & (GE_VTYPE_TC_MASK | GE_VTYPE_THROUGH_MASK))
-			gstate_c.Dirty(DIRTY_UVSCALEOFFSET);
-		// In this case, we may be doing weights and morphs.
-		// Update any bone matrix uniforms so it uses them correctly.
-		if ((op & GE_VTYPE_MORPHCOUNT_MASK) != 0) {
-			gstate_c.Dirty(gstate_c.deferredVertTypeDirty);
-			gstate_c.deferredVertTypeDirty = 0;
+		// If the vertex format changed, but the component setup didn't (same components, different input format),
+		// we don't actually need to flush. Assuming the output vertex format stays the same. This is quite common in games with highly optimized mesh data.
+		if ((diff & ~(GE_VTYPE_TC_MASK | GE_VTYPE_COL_MASK | GE_VTYPE_POS_MASK)) == 0 &&
```

Comment:
Couldn't we even ignore weightcount here?  I feel like this logic would be easier to read in a static func, i.e.:

```c++
static bool IsVTypeCompatible(u32 prev, u32 diff) {
	// Did anything outside the component types and weightcount change?
	if ((diff & ~(GE_VTYPE_WEIGHTCOUNT_MASK | GE_VTYPE_TC_MASK | GE_VTYPE_COL_MASK | GE_VTYPE_POS_MASK)) != 0)
		return false;

	u32 cur = prev ^ diff;
	if (((prev & GE_VTYPE_TC_MASK) != 0) != ((cur & GE_VTYPE_TC_MASK) != 0))
		return false;
	if (((prev & GE_VTYPE_COL_MASK) != 0) != ((cur & GE_VTYPE_COL_MASK) != 0))
		return false;

	return true;
}
```

Generally, I hate ifs that span multiple lines.  They're never clean, readable code.

Also, can't we actually ignore morphcount too?  It only affects vertex decode.  Might help LBP.

-[Unknown]

Reply:
Yeah we can definitely ignore morphcount here too, since changing it doesn't change the output format. Just didn't think of it. And breaking the function out like that makes sense too.